route-rs-packets/src/types.rs

```diff
@@ -0,0 +1,336 @@
+// Let's use this area for now to declare common structs, constants, and common helper functions.
+use std::fmt;
+
+/// The common datatype that all packet structures share to repreasent their data
+pub type PacketData<'packet> = &'packet mut Vec<u8>;
+
+// Most significant byte is 0th
+#[derive(Eq, Clone, Copy, Hash, PartialEq, Debug)]
+pub struct MacAddr {
+    pub bytes: [u8; 6],
+}
+
+impl MacAddr {
+    pub fn new(bytes: [u8; 6]) -> MacAddr {
+        MacAddr { bytes }
+    }
+}
+
+impl fmt::Display for MacAddr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{:X}:{:X}:{:X}:{:X}:{:X}:{:X}",
+            self.bytes[0],
+            self.bytes[1],
+            self.bytes[2],
+            self.bytes[3],
+            self.bytes[4],
+            self.bytes[5]
+        )
+    }
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Eq, PartialEq, Debug)]
+pub enum IpProtocol {
+    HOPOPT,
+    ICMP,
+    IGMP,
+    GGP,
+    IP_in_IP,
+    ST,
+    TCP,
+    CMT,
+    EGP,
+    IGP,
+    BBN_RRC_MON,
+    NVP_II,
+    PUP,
+    ARGUS,
+    EMCON,
+    XNET,
+    CHAOS,
+    UDP,
+    MUX,
+    DCN_MEAS,
+    HMP,
+    PRM,
+    XNS_IDP,
+    TRUNK_1,
+    TRUNK_2,
+    LEAF_1,
+    LEAF_2,
+    RDP,
+    IRTP,
+    ISO_TP4,
+    NETBLT,
+    MFE_NSP,
+    MERIT_INP,
+    DCCP,
+    THREEPC,
+    IDPR,
+    XTP,
+    DDP,
+    IDPR_CMTP,
+    TP_PLUS_PLUS,
+    IL,
+    IPv6,
+    SDRP,
+    IPv6_route,
+    IPv6_frag,
+    IDRP,
+    RSVP,
+    GREs,
+    DSR,
+    BNA,
+    ESP,
+    AH,
+    I_NLSP,
+    SwIPe,
+    NARP,
+    MOBILE,
+    TLSP,
+    SKIP,
+    IPv6_ICMP,
+    IPv6_NoNxt,
+    IPv6_Opts,
+    any_host_internal_protocol,
+    CFTP,
+    local_network,
+    SAT_EXPAK,
+    KRYPTOLAN,
+    RVD,
+    IPPC,
+    any_distributed_file_system,
+    SAT_MON,
+    VISA,
+    IPCU,
+    CPNX,
+    CPHB,
+    WSN,
+    PVP,
+    BR_SAT_MON,
+    SUN_ND,
+    WB_MON,
+    WB_EXPAK,
+    ISO_IP,
+    VMTP,
+    SECURE_VMTP,
+    VINES,
+    TTP,
+    NSFNET_IGP,
+    DGP,
+    TCF,
+    EIGRP,
+    OSPF,
+    Sprite_RPC,
+    LARP,
+    MTP,
+    AX_25,
+    OS,
+    MICP,
+    SCC_SP,
+    ETHERIP,
+    ENCAP,
+    any_private_encryption_scheme,
+    GMTP,
+    IFMP,
+    PNNI,
+    PIM,
+    ARIS,
+    SCPS,
+    QNX,
+    A_N,
+    IPComp,
+    SNP,
+    Compaq_Peer,
+    IPX_in_IP,
+    VRRP,
+    PGM,
+    any_0_hop_protocol,
+    L2TP,
+    DDX,
+    IATP,
+    STP,
+    SRP,
+    UTI,
+    SMP,
+    SM,
+    PTP,
+    IS_IS_over_IPv4,
+    FIRE,
+    CRTP,
+    CRUDP,
+    SSCOPMCE,
+    IPLT,
+    SPS,
+    PIPE,
+    SCTP,
+    FC,
+    RSVP_E2E_IGNORE,
+    Mobility_Header,
+    UDPLite,
+    MPLS_in_IP,
+    manet,
+    HIP,
+    Shim6,
+    WESP,
+    ROHC,
+    Unassigned,
+    Use_for_expiramentation_and_testing,
+    Reserved,
+}
+
+impl From<u8> for IpProtocol {
```

Comment:
Couple thoughts:
- It feels wordy to have to do this twice
- How do we turn an IpProtocol into a u8 when writing out a packet/overwriting the protocol field?

Have you seen the solution in libpnet? https://github.com/libpnet/libpnet/blob/master/pnet_packet/src/ip.rs It looks like the protocol is a struct with a single u8 field, then there are two mappings, one a list of constants, and one a display function.

Reply:
Hmm this is a decent solution.  I'd be fine with changing it to something like libpnet's

Reply:
Yeah, arguably `From<T>` isn't what we want here -- let's make clients explicitly convert integers to the typed values.

Reply:
Going to shelve this for after merge cleanup.

Reply:
#178